### ISSUE TITLE
Update to use delegated-routing for querying storetheindex

### ISSUE THREAD

**Issue by BigLep** [libp2p/hydra-booster#162]
## Done Criteria
Hydras query storetheindex using ipfs/go-delegated-routing rather than the storetheindex provider that was added in https://github.com/libp2p/hydra-booster/pull/158

## Why Important
Provides first production validation of delegated routing, giving us confidence to add it to go-ipfs as part of https://github.com/ipfs/go-ipfs/issues/8775

## Notes
1. We will use the ipld/edelweiss generated version of ipfs/go-delegated-routing happening in https://github.com/ipfs/go-delegated-routing/pull/11
2. The depends on storetheindex to expose a delegated-routing endpoint, happening in https://github.com/filecoin-project/storetheindex/issues/251

**Comment by petar** [comment]
@BigLep All tasks are done here, specifically:
- Hydra has metrics for Reframe path
- STI has metrics for Reframe path
- All old STI code is removed from Hydra
- Both Hydra and STI use the newest version of Delegated Routing and Edelweiss
- We've verified that Hydra and STI talk to each other

The next steps would be:
- deployment of STI to production (@willscott), then
- deployment of Hydra to production (@petar)

**Comment by BigLep** [comment]
Thanks @petar .  Lets track the storetheindex production deployment in https://github.com/filecoin-project/storetheindex/issues/251 .  Hydra production deployment will be here.

To be clear, has https://protocollabs.grafana.net/d/lNGQTv9Zz/hydra-boosters?orgId=1&var-flight=hydra-test- been updated to use the go-delegated-routing metrics rather than the existing sti-provider metrics?

Good stuff - almost there!

**Comment by BigLep** [comment]
@petar : following up, has https://protocollabs.grafana.net/d/lNGQTv9Zz/hydra-boosters?orgId=1&var-flight=hydra-test- been updated to use the go-delegated-routing metrics rather than the existing sti-provider metrics?

**Comment by petar** [comment]
> @petar : following up, has https://protocollabs.grafana.net/d/lNGQTv9Zz/hydra-boosters?orgId=1&var-flight=hydra-test- been updated to use the go-delegated-routing metrics rather than the existing sti-provider metrics?

Yes. The delegated routing code replaces the STI code and uses the same metric names. So the dashboard should work unchanged.

**Comment by BigLep** [comment]
@petar : last thing for closing this out.  Has the custom storetheindex code in libp2p/hydra-booster been removed?

**Comment by petar** [comment]
> @petar : last thing for closing this out.  Has the custom storetheindex code in libp2p/hydra-booster been removed?

Yes.

**Comment by willscott** [comment]
@thattommyhall I believe you pinged that we're still using the older protocol for all but the test instance - and that's what i see on the dashboard as well.

I didn't see the removal of the http indexer code go past on github, but noting that it means we probably should coordinate broader deployment of reframe before we end up with a deployment that doesn't support the current setup.

**Comment by BigLep** [comment]
Discussion about this effort is currently happening in the #reframe channel: https://filecoinproject.slack.com/archives/C03RQFURZLM

**Comment by BigLep** [comment]
Per 2022-08-12 verbal conversations, @guseggert is going to drive this effort to close and will consult with @petar as needed.

**Comment by guseggert** [comment]
So the logs originally showed the timeouts were due to a timeout while reading the response body: 

```
2022-08-14T09:37:46.803Z	ERROR	service/client/delegatedrouting	proto/proto_edelweiss.go:1234	client received error response (context deadline exceeded (Client.Timeout or context cancellation while reading body))
```

This morning I deployed https://github.com/libp2p/hydra-booster/commit/37dda2209d91ad0dd213534248382e0f144d1991 to the test flight, which publishes more detailed error metrics for Reframe and also upgrades to go-libp2p@v0.21 and Go 1.18.  After deployment, the timeouts basically disappeared, and it's been baking for a few hours and the traffic is at similar levels now without timeouts. This leads me to believe that the issue is probably that the Hydra node was taking too long to read the response body due to some environmental issue (overload from some other work it was doing). Something in the go-libp2p or Go upgrades might have also alleviated the bottleneck, e.g. libp2p resource manager. We see similar timeouts in prod with the non-Reframe StoreTheIndex client, although not nearly at the same rate, but the test flight could have gotten unlucky and been placed into a hot partition, so it could still be the same issue.

My next step is to get some metrics into the dashboard on libp2p Resource Manager to see if it's throttling anything, and understand the impact of that on the network, see if we need to tweak limits, etc. I'm guessing that RM is throttling b/c the AddProvider rate is much lower, while the STI rate is the same.

Enumerating the options to mitigate overloading:

* Add a rate limiter to cap the rate of AddProvider DHT calls
  * We should do this regardless of the other options, as this is the only way we can prevent nodes from being overloaded when traffic patterns change
  * Some calls will start failing for the other nodes that are calling AddProvider, what's the impact of this?
  * This might already be happening with the libp2p upgrade and libp2p resource manager
* Reduce the number of heads that each node runs
  * This will increase the overall cost as we'll need to scale up the fleet to accomodate
  * Traffic pattern changes in the network could still cause overloads
* Do some analysis on the nature of the calls to see if caching could alleviate the load
  * Are there hot CIDs/addrs that we could shed w/ caching?
  * Traffic pattern changes in the network could still cause overloads

**Comment by guseggert** [comment]
I've integrated Resource Manager, added RM metrics, added them to the dashboard, and tweaked the RM limits to be low enough to throttle spikes but to generally allow most traffic. The test node is now operating at the same capacity as before, but with minimal timeouts. There are still occasional timeouts (about 0.3% of reqs). These are timeouts reading response headers, so this may be a server-side thing, although I will increase the client-side timeout to 3 seconds to allow for e.g. GC to run w/o causing timeouts.

I'm working on this branch: https://github.com/libp2p/hydra-booster/tree/feat/reframe-metrics

I'll get a PR worked up, and continue to let this bake today. If it looks okay tomorrow morning, I'll roll it out to the rest of the fleet.

**Comment by BigLep** [comment]
2022-08-19:
* We've deployed to the test suite.
* PR is out with updated libp2p, go version, metrics, etc.
* Planning to deploy to production 2022-08-22

**Comment by BigLep** [comment]
@guseggert : other thoughts from looking at this after:
1. I worry that it isn't going to be clear for anyone looking at "StoreTheIndex Reuests / Sec" what "Net", "NetTimeout", and "Other" mean.  Can we maybe add an "info panel" (assuming something like that exists) with an explainer note and link to canonical information.
2. Please handle https://github.com/ipfs/specs/issues/308 and ensure go-delegated-routing is doing the right thing.
3. For the latency metrics, do we have other values besides average.  For example, I'm curious what the p99/p100 is for "success". 
4. Did we do this from the done criteria: "Distribution of 2xx response payloads sizes (in terms of number of records). For each 2xx responses, we should accumulate a metric for the number of providers in the response. This allows us to say the the p90 of responses have X providers."

<img width="1365" alt="image" src="https://user-images.githubusercontent.com/85411/185720758-28646d01-4370-4ba1-877b-1fa07d10868c.png">

**Comment by guseggert** [comment]
1) is done.

I did 2) a couple weeks ago, see https://github.com/ipfs/specs/commit/0b123fa5c3ddf126f55bcec2aa67b2b91247d959

3) done in https://github.com/libp2p/hydra-booster/commit/6967a65c3f6945c40a6cc1ac8d378bbf173e539d

4) done in https://github.com/libp2p/hydra-booster/commit/d48d8984c19c125c55c199a5e790bed93d9ed2b7

**Comment by guseggert** [comment]
Update: last week I deployed Reframe to the full Hydra fleet, but almost all reqs started timing out so I rolled it back. Have been debugging w/ @masih in between traveling.

Yesterday there was an STI event that caused the HTTP endpoint to behave a like the Reframe timeouts, so I'm working with @masih to understand the root cause. If it doesn't rule out Reframe, then I'll wait for the root cause fix and redeploy to see if it also works for Reframe, and if it doesn't then we'll need to do some req tracing through the infrastructure to see where exactly the timeouts are occurring. This might require adding request IDs to the reqs and passing those through LBs, proxies, etc., and adding to log messages on the STI side.

**Comment by masih** [comment]
All fixes for the STI outage yesterday are now deployed. At this time it is unclear if the fixes would also resolve the timeouts observed when reframe was deployed. We try deployment and see if it does if that's not too disruptive to the users in case it doesn't.

**Comment by BigLep** [comment]
Thanks for the updates guys.  I'll keep following - let me know if anything is needed.

**Comment by petar** [comment]
Yes.
